File: app/src/main/java/com/colormixlab/game/GameViewModel.kt
package com.colormixlab.game

import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.colormixlab.model.GameColor
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

/**
 * Android-specific ViewModel - thin wrapper around shared GameController
 * Only handles platform-specific concerns (timer coroutines, Compose State conversion)
 */
class GameViewModel : ViewModel() {
    // Shared game controller contains ALL game logic
    private val gameController = GameController()

    // Convert StateFlow to Compose State for UI observation
    private val _gameState = mutableStateOf(gameController.gameState.value)
    val gameState: State<GameState> = _gameState

    private var timerJob: Job? = null

    init {
        // Observe shared game state and convert to Compose State
        gameController.gameState.onEach { newState ->
            _gameState.value = newState

            // Handle timer lifecycle based on state changes
            if (newState.isTimerActive && timerJob == null) {
                startTimerCoroutine()
            } else if (!newState.isTimerActive && timerJob != null) {
                cancelTimerCoroutine()
            }
        }.launchIn(viewModelScope)
    }

    // ========== DELEGATE ALL ACTIONS TO SHARED CONTROLLER ==========

    fun setDifficulty(difficulty: Difficulty) {
        gameController.setDifficulty(difficulty)
    }

    fun addColorDrop(color: GameColor) {
        gameController.addColorDrop(color)
    }

    fun clearBowl() {
        gameController.clearBowl()
    }

    fun checkMatch() {
        gameController.checkMatch()
    }

    fun calculatePoints(similarity: Float): Int {
        return gameController.calculatePoints(similarity)
    }

    fun getResultMessage(similarity: Float): String {
        return gameController.getResultMessage(similarity)
    }

    fun getResultEmoji(similarity: Float): String {
        return gameController.getResultEmoji(similarity)
    }

    fun nextLevel() {
        gameController.nextLevel()
    }

    fun retryLevel() {
        gameController.retryLevel()
    }

    fun dismissSuccessDialog() {
        _gameState.value = _gameState.value.copy(showSuccessDialog = false)
    }

    fun completeMathChallenge() {
        gameController.completeMathChallenge()
    }

    fun getCurrentMathChallenge(): Pair<MathChallengeType, GameColor?>? {
        return gameController.getCurrentMathChallenge()
    }

    fun deductPointsForWrongMathAnswer() {
        gameController.deductPointsForWrongMathAnswer()
    }

    fun exitGame() {
        cancelTimerCoroutine()
    }

    fun resetGame() {
        cancelTimerCoroutine()
        gameController.resetGame()
    }

    fun completeGame() {
        _gameState.value = _gameState.value.copy(isGameCompleted = false)
    }

    fun forceFinishGame() {
        cancelTimerCoroutine()
        gameController.forceFinishGame()
    }

    // ========== ANDROID-SPECIFIC TIMER MANAGEMENT ==========
    // Timer runs in viewModelScope coroutine and calls shared controller's tickTimer()

    private fun startTimerCoroutine() {
        timerJob?.cancel()
        timerJob = viewModelScope.launch {
            while (true) {
                delay(1000)
                gameController.tickTimer()

                // Check if timer is still active
                if (!gameController.gameState.value.isTimerActive) {
                    break
                }
            }
        }
    }

    fun pauseTimer() {
        gameController.pauseTimer()
    }

    fun resumeTimer() {
        gameController.resumeTimer()
    }

    private fun cancelTimerCoroutine() {
        timerJob?.cancel()
        timerJob = null
    }
}

File: iosApp/ColorMixLab/ColorMixLab/GameViewModel.swift
//
//  GameViewModel.swift
//  ColorMixLab iOS
//
//  iOS-specific ViewModel - thin wrapper around shared GameController
//  Only handles platform-specific concerns (timer, haptics, color conversions)
//

import Foundation
import SwiftUI
import Combine
import shared

class GameViewModel: ObservableObject {
    @Published var gameState: GameState

    // Shared game controller contains ALL game logic
    private let gameController: GameController

    private var timerCancellable: AnyCancellable?
    private var stateObserver: Cancellable?
    private let hapticProvider = HapticProvider()

    init(difficulty: Difficulty) {
        // Create shared controller
        gameController = GameController(initialDifficulty: difficulty)
        gameState = gameController.gameState.value as! GameState

        // Kotlin StateFlow observation from Swift is awkward, so we poll the
        // current value on a 100ms cadence and republish when it changes.
        // Change detection uses the shared data class's generated equality.
        stateObserver = Timer.publish(every: 0.1, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                guard let self = self else { return }
                guard let newState = self.gameController.gameState.value as? GameState else { return }
                if !newState.isEqual(self.gameState) {
                    self.gameState = newState

                    // Handle timer lifecycle based on state changes
                    if newState.isTimerActive && self.timerCancellable == nil {
                        self.startTimer()
                    } else if !newState.isTimerActive && self.timerCancellable != nil {
                        self.stopTimer()
                    }
                }
            }
    }

    // MARK: - Actions (delegated to shared controller)

    func addColorDrop(color: GameColor) {
        gameController.addColorDrop(color: color)
        hapticProvider.performHaptic(type: HapticType.lightTap)
    }

    func clearBowl() {
        gameController.clearBowl()
        hapticProvider.performHaptic(type: HapticType.lightTap)
    }

    func checkMatch() {
        gameController.checkMatch()
        let isSuccess = gameState.similarity >= GameConstants.shared.MATCH_SUCCESS_THRESHOLD
        hapticProvider.performHaptic(type: isSuccess ? HapticType.success : HapticType.error)
    }

    func nextLevel() {
        gameController.nextLevel()
    }

    func retryLevel() {
        gameController.retryLevel()
    }

    func resetGame() {
        stopTimer()
        gameController.resetGame()
    }

    func completeMathChallenge() {
        gameController.completeMathChallenge()
    }

    func getCurrentMathChallenge() -> (MathChallengeType, GameColor?)? {
        guard let result = gameController.getCurrentMathChallenge() else { return nil }
        return (result.first as! MathChallengeType, result.second as? GameColor)
    }

    func deductPointsForWrongAnswer() {
        gameController.deductPointsForWrongMathAnswer()
    }

    func exitGame() {
        stopTimer()
        stateObserver?.cancel()
    }

    func resultMessage(for similarity: Float) -> String {
        gameController.getResultMessage(similarity: similarity)
    }

    func getResultEmoji(similarity: Float) -> String {
        gameController.getResultEmoji(similarity: similarity)
    }

    // MARK: - Timer management
    // Timer runs on main thread and calls shared controller's tickTimer()

    private func startTimer() {
        timerCancellable?.cancel()

        timerCancellable = Timer.publish(every: 1, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                self?.gameController.tickTimer()
            }
    }

    private func stopTimer() {
        timerCancellable?.cancel()
        timerCancellable = nil
    }

    // MARK: - UI color conversions

    var targetColor: Color {
        gameState.targetColor.swiftUIColor
    }

    var mixedColor: Color {
        gameState.mixedColor.swiftUIColor
    }
}
